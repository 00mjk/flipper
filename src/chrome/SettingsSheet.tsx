/**
 * Copyright 2018-present Facebook.
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * @format
 */

import {FlexColumn, Button, styled, Text, FlexRow, Spacer} from 'flipper';
import React, {Component} from 'react';
import {updateSettings, Action} from '../reducers/settings';
import {connect} from 'react-redux';
import {State as Store} from '../reducers';
import {Settings} from '../reducers/settings';
import {flush} from '../utils/persistor';
import ToggledSection from './settings/ToggledSection';
import {FilePathConfigField} from './settings/configFields';
import {remote} from 'electron';
import isEqual from 'lodash.isequal';

const Container = styled(FlexColumn)({
  padding: 20,
  width: 800,
});

const Title = styled(Text)({
  marginBottom: 18,
  marginRight: 10,
  fontWeight: 100,
  fontSize: '40px',
});

type OwnProps = {
  onHide: () => void;
};

type StateFromProps = {
  settings: Settings;
};

type DispatchFromProps = {
  updateSettings: (settings: Settings) => Action;
};

type State = {
  updatedSettings: Settings;
};

type Props = OwnProps & StateFromProps & DispatchFromProps;
class SettingsSheet extends Component<Props, State> {
  state: State = {
    updatedSettings: {...this.props.settings},
  };

  applyChanges = async () => {
    this.props.updateSettings(this.state.updatedSettings);
    this.props.onHide();
    flush().then(() => {
      remote.app.relaunch();
      remote.app.exit();
    });
  };

  render() {
    return (
      <Container>
        <Title>Settings</Title>
        <ToggledSection
          label="Android"
          enabled={this.state.updatedSettings.enableAndroid}
          onChange={v => {
            this.setState({
              updatedSettings: {
                ...this.state.updatedSettings,
                enableAndroid: v,
              },
            });
          }}>
          <FilePathConfigField
            label="Android SDK Location"
            defaultValue={this.state.updatedSettings.androidHome}
            onChange={v => {
              this.setState({
                updatedSettings: {
                  ...this.state.updatedSettings,
                  androidHome: v,
                },
              });
            }}
          />
        </ToggledSection>

        <br />
        <FlexRow>
          <Spacer />
          <Button compact padded onClick={this.props.onHide}>
            Cancel
          </Button>
          <Button
            disabled={isEqual(this.props.settings, this.state.updatedSettings)}
            type="primary"
            compact
            padded
            onClick={this.applyChanges}>
            Apply and Restart
          </Button>
        </FlexRow>
      </Container>
    );
  }
}

export default connect<StateFromProps, DispatchFromProps, OwnProps, Store>(
  ({settingsState}) => ({settings: settingsState}),
  {updateSettings},
)(SettingsSheet);
