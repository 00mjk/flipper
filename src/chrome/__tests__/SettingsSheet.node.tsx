/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import SettingsSheet from '../SettingsSheet';
import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';

function getStore({
  iosEnabled,
  androidEnabled,
  androidHome,
}: {
  iosEnabled?: boolean;
  androidEnabled?: boolean;
  androidHome?: string;
}) {
  return configureStore([])({
    application: {
      sessionId: 'mysession',
      xcodeCommandLineToolsDetected: iosEnabled,
    },
    settingsState: {
      enableAndroid: androidEnabled,
      androidHome: androidHome,
    },
  });
}

test('SettingsSheet snapshot with nothing enabled', () => {
  const component = (
    <Provider store={getStore({})}>
      <SettingsSheet onHide={() => {}} />
    </Provider>
  );

  expect(renderer.create(component).toJSON()).toMatchSnapshot();
});

test('SettingsSheet snapshot with iOS enabled', () => {
  const component = (
    <Provider store={getStore({iosEnabled: true})}>
      <SettingsSheet onHide={() => {}} />
    </Provider>
  );

  expect(renderer.create(component).toJSON()).toMatchSnapshot();
});

test('SettingsSheet snapshot with android enabled', () => {
  const component = (
    <Provider store={getStore({androidEnabled: true})}>
      <SettingsSheet onHide={() => {}} />
    </Provider>
  );

  expect(renderer.create(component).toJSON()).toMatchSnapshot();
});

test('SettingsSheet snapshot with android home set', () => {
  const component = (
    <Provider store={getStore({androidHome: '/my/sdk/location'})}>
      <SettingsSheet onHide={() => {}} />
    </Provider>
  );

  expect(renderer.create(component).toJSON()).toMatchSnapshot();
});
