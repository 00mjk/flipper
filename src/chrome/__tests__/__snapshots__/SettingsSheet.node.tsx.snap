// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsSheet snapshot with android enabled 1`] = `
<div
  className="css-58lehk"
>
  <span
    className="css-1lhexul"
  >
    Settings
  </span>
  <div
    className="css-1k4677w"
  >
    <div
      className="css-a1glw"
    >
      <div
        className="css-xu6aw0"
        onClick={[Function]}
      >
        <div
          className="css-5fb050"
        />
        <span
          className="css-1ppqprl"
        >
          Android Developer
        </span>
      </div>
    </div>
    <div
      className="css-77h5wk"
    >
      <div
        className="css-1qb4gb9"
      >
        <span
          className="css-1244i8h"
        >
          Android SDK Location
        </span>
        <input
          className="css-g76x1o"
          onChange={[Function]}
          placeholder="Android SDK Location"
          type="text"
        />
        <div
          className="css-1k4677w"
          onClick={[Function]}
        >
          <img
            alt="dots-3-circle"
            className="css-1x9cv9"
            size={16}
            src="https://external.xx.fbcdn.net/assets/?name=dots-3-circle&variant=outline&size=16&set=facebook_icons&density=1x"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="css-1k4677w"
  >
    <div
      className="css-a1glw"
    >
      <div
        className="css-xu6aw0"
        onClick={[Function]}
      >
        <div
          className="css-1bpd2m8"
        />
        <span
          className="css-1ppqprl"
        >
          iOS Developer
        </span>
      </div>
      <div
        className="css-ivsi5g"
      />
    </div>
    <div
      className="css-77h5wk"
    >
       
      <div
        className="css-1qb4gb9"
      >
        <span
          className="css-1244i8h"
        >
          Use xcode-select to enable or switch between xcode versions
        </span>
        <div
          className="css-ivsi5g"
        />
      </div>
      <div
        className="css-ivsi5g"
      />
    </div>
  </div>
  <br />
  <div
    className="css-a1glw"
  >
    <div
      className="css-12zzrdt"
    />
    <div
      className="css-1awwm1c"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      Cancel
    </div>
    <div
      className="css-ytyipw"
      disabled={true}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="primary"
    >
      Apply and Restart
    </div>
  </div>
</div>
`;

exports[`SettingsSheet snapshot with android home set 1`] = `
<div
  className="css-58lehk"
>
  <span
    className="css-1lhexul"
  >
    Settings
  </span>
  <div
    className="css-1k4677w"
  >
    <div
      className="css-a1glw"
    >
      <div
        className="css-xu6aw0"
        onClick={[Function]}
      >
        <div
          className="css-1bpd2m8"
        />
        <span
          className="css-1ppqprl"
        >
          Android Developer
        </span>
      </div>
    </div>
    <div
      className="css-77h5wk"
    >
      <div
        className="css-1qb4gb9"
      >
        <span
          className="css-1244i8h"
        >
          Android SDK Location
        </span>
        <input
          className="css-g76x1o"
          onChange={[Function]}
          placeholder="Android SDK Location"
          type="text"
          value="/my/sdk/location"
        />
        <div
          className="css-1k4677w"
          onClick={[Function]}
        >
          <img
            alt="dots-3-circle"
            className="css-1x9cv9"
            size={16}
            src="https://external.xx.fbcdn.net/assets/?name=dots-3-circle&variant=outline&size=16&set=facebook_icons&density=1x"
          />
        </div>
      </div>
      <div
        className="css-ivsi5g"
      />
    </div>
  </div>
  <div
    className="css-1k4677w"
  >
    <div
      className="css-a1glw"
    >
      <div
        className="css-xu6aw0"
        onClick={[Function]}
      >
        <div
          className="css-1bpd2m8"
        />
        <span
          className="css-1ppqprl"
        >
          iOS Developer
        </span>
      </div>
      <div
        className="css-ivsi5g"
      />
    </div>
    <div
      className="css-77h5wk"
    >
       
      <div
        className="css-1qb4gb9"
      >
        <span
          className="css-1244i8h"
        >
          Use xcode-select to enable or switch between xcode versions
        </span>
        <div
          className="css-ivsi5g"
        />
      </div>
      <div
        className="css-ivsi5g"
      />
    </div>
  </div>
  <br />
  <div
    className="css-a1glw"
  >
    <div
      className="css-12zzrdt"
    />
    <div
      className="css-1awwm1c"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      Cancel
    </div>
    <div
      className="css-ytyipw"
      disabled={true}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="primary"
    >
      Apply and Restart
    </div>
  </div>
</div>
`;

exports[`SettingsSheet snapshot with iOS enabled 1`] = `
<div
  className="css-58lehk"
>
  <span
    className="css-1lhexul"
  >
    Settings
  </span>
  <div
    className="css-1k4677w"
  >
    <div
      className="css-a1glw"
    >
      <div
        className="css-xu6aw0"
        onClick={[Function]}
      >
        <div
          className="css-1bpd2m8"
        />
        <span
          className="css-1ppqprl"
        >
          Android Developer
        </span>
      </div>
    </div>
    <div
      className="css-77h5wk"
    >
      <div
        className="css-1qb4gb9"
      >
        <span
          className="css-1244i8h"
        >
          Android SDK Location
        </span>
        <input
          className="css-g76x1o"
          onChange={[Function]}
          placeholder="Android SDK Location"
          type="text"
        />
        <div
          className="css-1k4677w"
          onClick={[Function]}
        >
          <img
            alt="dots-3-circle"
            className="css-1x9cv9"
            size={16}
            src="https://external.xx.fbcdn.net/assets/?name=dots-3-circle&variant=outline&size=16&set=facebook_icons&density=1x"
          />
        </div>
      </div>
      <div
        className="css-ivsi5g"
      />
    </div>
  </div>
  <div
    className="css-1k4677w"
  >
    <div
      className="css-a1glw"
    >
      <div
        className="css-xu6aw0"
        onClick={[Function]}
      >
        <div
          className="css-5fb050"
        />
        <span
          className="css-1ppqprl"
        >
          iOS Developer
        </span>
      </div>
      <div
        className="css-ivsi5g"
      />
    </div>
    <div
      className="css-77h5wk"
    >
       
      <div
        className="css-1qb4gb9"
      >
        <span
          className="css-1244i8h"
        >
          Use xcode-select to enable or switch between xcode versions
        </span>
        <div
          className="css-ivsi5g"
        />
      </div>
    </div>
  </div>
  <br />
  <div
    className="css-a1glw"
  >
    <div
      className="css-12zzrdt"
    />
    <div
      className="css-1awwm1c"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      Cancel
    </div>
    <div
      className="css-ytyipw"
      disabled={true}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="primary"
    >
      Apply and Restart
    </div>
  </div>
</div>
`;

exports[`SettingsSheet snapshot with nothing enabled 1`] = `
<div
  className="css-58lehk"
>
  <span
    className="css-1lhexul"
  >
    Settings
  </span>
  <div
    className="css-1k4677w"
  >
    <div
      className="css-a1glw"
    >
      <div
        className="css-xu6aw0"
        onClick={[Function]}
      >
        <div
          className="css-1bpd2m8"
        />
        <span
          className="css-1ppqprl"
        >
          Android Developer
        </span>
      </div>
    </div>
    <div
      className="css-77h5wk"
    >
      <div
        className="css-1qb4gb9"
      >
        <span
          className="css-1244i8h"
        >
          Android SDK Location
        </span>
        <input
          className="css-g76x1o"
          onChange={[Function]}
          placeholder="Android SDK Location"
          type="text"
        />
        <div
          className="css-1k4677w"
          onClick={[Function]}
        >
          <img
            alt="dots-3-circle"
            className="css-1x9cv9"
            size={16}
            src="https://external.xx.fbcdn.net/assets/?name=dots-3-circle&variant=outline&size=16&set=facebook_icons&density=1x"
          />
        </div>
      </div>
      <div
        className="css-ivsi5g"
      />
    </div>
  </div>
  <div
    className="css-1k4677w"
  >
    <div
      className="css-a1glw"
    >
      <div
        className="css-xu6aw0"
        onClick={[Function]}
      >
        <div
          className="css-1bpd2m8"
        />
        <span
          className="css-1ppqprl"
        >
          iOS Developer
        </span>
      </div>
      <div
        className="css-ivsi5g"
      />
    </div>
    <div
      className="css-77h5wk"
    >
       
      <div
        className="css-1qb4gb9"
      >
        <span
          className="css-1244i8h"
        >
          Use xcode-select to enable or switch between xcode versions
        </span>
        <div
          className="css-ivsi5g"
        />
      </div>
      <div
        className="css-ivsi5g"
      />
    </div>
  </div>
  <br />
  <div
    className="css-a1glw"
  >
    <div
      className="css-12zzrdt"
    />
    <div
      className="css-1awwm1c"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      Cancel
    </div>
    <div
      className="css-ytyipw"
      disabled={true}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="primary"
    >
      Apply and Restart
    </div>
  </div>
</div>
`;
